azurerm/internal/services/kusto/kusto_cluster_principal_assignment_resource.go

```diff
@@ -0,0 +1,235 @@
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRMKustoClusterName,
+			},
+
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateAzureRMKustoClusterPrincipalAssignmentName,
+			},
+
+			"tenant_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"tenant_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"principal_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"principal_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"principal_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(kusto.PrincipalTypeApp),
+					string(kusto.PrincipalTypeGroup),
+					string(kusto.PrincipalTypeUser),
+				}, false),
+			},
+
+			"role": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(kusto.AllDatabasesAdmin),
+					string(kusto.AllDatabasesViewer),
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceArmKustoClusterPrincipalAssignmentCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Kusto.ClusterPrincipalAssignmentsClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for Azure Kusto Cluster Principal Assignment creation.")
+
+	resourceGroup := d.Get("resource_group_name").(string)
+	clusterName := d.Get("cluster_name").(string)
+	name := d.Get("name").(string)
+
+	if d.IsNewResource() {
+		principalAssignment, err := client.Get(ctx, resourceGroup, clusterName, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(principalAssignment.Response) {
+				return fmt.Errorf("Error checking for presence of existing Kusto Cluster Principal Assignment %q (Resource Group %q, Cluster %q): %+v", name, resourceGroup, clusterName, err)
+			}
+		}
+
+		if principalAssignment.ID != nil && *principalAssignment.ID != "" {
+			return tf.ImportAsExistsError("azurerm_kusto_cluster_principal_assignment", *principalAssignment.ID)
+		}
+	}
+
+	tenantID := d.Get("tenant_id").(string)
+	principalID := d.Get("principal_id").(string)
+	principalType := d.Get("principal_type").(string)
+	role := d.Get("role").(string)
+
+	props := kusto.ClusterPrincipalProperties{
+		TenantID:      utils.String(tenantID),
+		PrincipalID:   utils.String(principalID),
+		PrincipalType: kusto.PrincipalType(principalType),
+		Role:          kusto.ClusterPrincipalRole(role),
+	}
+
+	principalAssignment := kusto.ClusterPrincipalAssignment{
+		ClusterPrincipalProperties: &props,
+	}
```

Comment:
We could consolidate this quite a bit by assigning everything in `principalAssignment`